Ensure that MRP logic is not executed for messages over other Transports

pidarped opened the pull request:
MRP based Ack handling and retransmissions should not be performed when messages are not sent using MRP, e.g., over TCP.
Install guards on Send/Receive paths to steer traffic to bypass MRP logic when going over TCP.

Fixes #30104

bzbarsky-apple commented on src/messaging/ExchangeContext.cpp:
```diff
@@ -143,7 +143,7 @@ CHIP_ERROR ExchangeContext::SendMessage(Protocols::Id protocolId, uint8_t msgTyp
     {
         // ExchangeContext for group are supposed to always be Initiator
-        if (IsGroupExchangeContext() && !IsInitiator())
+        if (IsGroupExchangeContext() && (!IsInitiator() || GetSessionHandle()->RequireTCP()))
```
Please update the comment to explain why this check is what it is.

Because I think this check is wrong.  `GetSessionHandle()->RequireTCP()` is always false, as implemented, for the `IsGroupExchangeContext()` case.

pidarped replied:
Will update the comment. Is RequireTCP() guaranteed to be false in this case, always? The reason I put it was basically to ensure that, i.e, if it is GroupExchangeContext, then it should not try to use TCP.

bzbarsky-apple replied:
IsGroupExchangeContext() checks for a group session.  Group sessions return false from RequireTCP(), no?

bzbarsky-apple commented on src/messaging/ExchangeContext.cpp:
```diff
@@ -531,57 +531,60 @@ CHIP_ERROR ExchangeContext::HandleMessage(uint32_t messageCounter, const Payload
-        return CHIP_NO_ERROR;
-    }
+        // Since the message is duplicate, let's not forward it up the stack
+        if (isDuplicate)
```
This is wrong.  The duplicate check needs to be independent of MRP.

In general, it's worth thinking a bit about which of these parts should be conditioned on what.  For example, `IsAckPending()` is going to be false if the transport does not support MRP, so there's no reason to put it inside the check...  On the other hand, the question of what to do if the other side asks for an ack is an interesting one.  Right now I guess we will ack things over BTP if the other side asks us to...

Similarly, given the ExchangeManager changes, IsEphemeralExchange() will always be false in the TCP case, right?

pidarped replied:
Yes, you are right. Thanks. I was careful enough to keep duplicate handling out of this at the ExchangeMgr level but overlooked this one in ExchangeContext. Will take that check out.
On the other side asking for an Ack, I was thinking that it should not do so given how we structure the code on both sender and receiver side. There is no point in that over TCP and allowing that would necessarily mean allowing all of MRP over TCP which is a big waste of resources, right? 
IsEphemeralExchange() should also be false over TCP. I don't see any use for it over TCP. But it can stay inside since we do not want to check it for all TCP messages.

bzbarsky-apple replied:
> I was thinking that it should not do so

If the other side is the Matter SDK, absolutely.

If the other side is just sending various things, we just need to handle it somewhat gracefully.  How gracefully is up to us.

> IsEphemeralExchange() should also be false over TCP

Right, so we don't have to put it under the "is tcp" bit.

My main point is this: TCP is not special.  Whatever is happening for TCP should happen for BTP too.  Our existing code should generally all work with this, and if it does not, we should fix those bugs... But the fixes should not be TCP-specific.

pidarped replied:
Agreed that TCP and BTP should follow similar flows and both should essentially exclude MRP retransmission logic. Not sure about the other side sending various things, but the receiving stack should ignore and bypass all MRP logic when handling messages over connection-oriented transports. If the other side is not Matter, and it is requesting an MRP Ack over TCP(by whatever means), the local side need not oblige and send back any acknowledgments, right?

>Right, so we don't have to put it under the "is tcp" bit.

Do you mean under the AllowsMRP() check? I have put it outside currently. But I was of the opinion that all these checks are relevant to only MRP traffic and TCP|BTP messages should not need to go through these checks. From that standpoint, only IsDuplicate() should be outside, isn't it?

pidarped replied:
Agreed that TCP and BTP should follow similar flows and both should essentially exclude MRP retransmission logic. Not sure about the other side sending various things, but the receiving stack should ignore and bypass all MRP logic when handling messages over connection-oriented transports. If the other side is not Matter, and it is requesting an MRP Ack over TCP(by whatever means), the local side need not oblige and send back any acknowledgments, right?

>Right, so we don't have to put it under the "is tcp" bit.

Do you mean under the AllowsMRP() check? I have put it outside currently. But I was of the opinion that all these checks are relevant to only MRP traffic and TCP|BTP messages should not need to go through these checks. From that standpoint, only IsDuplicate() should be outside, isn't it?

bzbarsky-apple replied:
> Do you mean under the AllowsMRP() check? 

Right.

I don't feel too strongly about the ephemeral exchange check; if the code is simpler with it inside the `AllowsMRP()` block, great. @pidarped

bzbarsky-apple commented on src/messaging/ExchangeMessageDispatch.cpp:
```diff
@@ -51,45 +51,60 @@ CHIP_ERROR ExchangeMessageDispatch::SendMessage(SessionManager * sessionManager,
 
-    // If there is a pending acknowledgment piggyback it on this message.
-    if (reliableMessageContext->HasPiggybackAckPending())
+    if (session->RequireTCP())
```
I don't understand why all the surgery here.  `reliableMessageContext->AutoRequestAck()` should be returning the right thing for TCP, and then the existing code would just work, no?

In any case, there shouldn't be separate cases for handling BTP and TCP here; they should take the same codepath.

pidarped replied:
I hadn't paid close attention to that, but I think we can use it. Currently, it wouldn't work though as AutoRequestAck is set for all non-group sessions I think. So, if we set it for only UDP based sessions, then the existing logic should work for TCP.

pidarped replied:
There is also the HasPiggybackAckPending() check. There are all these checks that are only relevant for MRP and should be avoided entirely for TCP messages instead of making TCP messages pass through them(even if false) I think.

bzbarsky-apple replied:
> as AutoRequestAck is set for all non-group sessions I think

OK, but then how do things work for BTP right now?  The answer is presumably that either they are buggy or they are already gated on the misnamed `RequireMRP()` thing.

But again: TCP is not special.  BTP and TCP should take the same codepaths.

I am fine with making that codepath clearer by avoiding calling functions, even if those functions would be guaranteed no-ops because they do their own checks.  But that should not be conditioned on "tcp".  It should be conditioned on whether the session allows MRP.

pidarped replied:
So, it seems AutoRequestAck is set in the ExchangeContext ctor. Do we currently use MRP over BLE? We should not, right? 
>It should be conditioned on whether the session allows MRP.

Yes, that is how it is in the latest change.

bzbarsky-apple replied:
> Do we currently use MRP over BLE?

We shouldn't, and I am pretty sure we do not.

pidarped replied:
That is what I would expect too. But it is set for BLE message in the ExchangeContext ctor. So, somewhere down the path, I guess it gets overridden by the RequireMRP() check, which is false for BLE.

bzbarsky-apple commented on src/messaging/ExchangeMgr.cpp:
```diff
@@ -352,7 +352,9 @@ void ExchangeManager::OnMessageReceived(const PacketHeader & packetHeader, const
             ChipLogError(ExchangeManager, "OnMessageReceived failed, err = %" CHIP_ERROR_FORMAT,
                          CHIP_ERROR_INVALID_MESSAGE_TYPE.Format());
             ec->Close();
+
```
Why the random whitespace changes?

pidarped replied:
Will remove. I may have added and deleted something earlier and these lingered.

bzbarsky-apple commented on src/transport/SecureSession.h:
```diff
@@ -158,6 +158,8 @@ class SecureSession : public Session, public ReferenceCounted<SecureSession, Sec
 
     bool RequireMRP() const override { return GetPeerAddress().GetTransportType() == Transport::Type::kUdp; }
```
Pre-existing, but this makes no sense.  UDP _allows_ MRP, but does not require it.  Follow-up issue for fixing that?

But fundamentally, I don't think we need a `RequireTCP` function.  We should have `AllowsMRP` (false for group sessions, true for unicast UDP sessions, false for unicast BTP/TCP sessions; it's the current RequireMRP, but named clearly), and just test that as needed.

pidarped replied:
Yes, I was also questioning the naming and just went with the existing model. I will change it to AllowsMRP() in this PR which makes more sense. I want to have a RequiresTCP() as I was planning to use it for other purposes on the session, as an easy way to determine what approach to adopt on the session handling side.

bzbarsky-apple replied:
That's fine, but none of the code being changed for this PR should need to use RequiresTCP.

And maybe RequiresTCP should be called TransportIsTCP() or something?

pidarped replied:
Made it IsTransportTCP()

bzbarsky-apple commented on src/messaging/ExchangeContext.cpp:
```diff
@@ -113,7 +113,7 @@ CHIP_ERROR ExchangeContext::SendMessage(Protocols::Id protocolId, uint8_t msgTyp
     // transmission.
     bool reliableTransmissionRequested =
-        GetSessionHandle()->RequireMRP() && !sendFlags.Has(SendMessageFlags::kNoAutoRequestAck) && !IsGroupExchangeContext();
+        GetSessionHandle()->AllowsMRP() && !sendFlags.Has(SendMessageFlags::kNoAutoRequestAck) && !IsGroupExchangeContext();
```
The IsGroupExchangeContext() check here is redundant, right?  Pre-existing, so maybe just file a followup issue.

pidarped replied:
Yes, It is redundant. I didn't realize that earlier given how the checks are structured. Will leave them here for now but will file an issue.

bzbarsky-apple commented on src/messaging/ExchangeContext.cpp:
```diff
@@ -142,8 +142,9 @@ CHIP_ERROR ExchangeContext::SendMessage(Protocols::Id protocolId, uint8_t msgTyp
-        if (IsGroupExchangeContext() && !IsInitiator())
+        // ExchangeContext for group are supposed to always be Initiator AND not
+        // use TCP as the underlying transport.
+        if (IsGroupExchangeContext() && (!IsInitiator() || GetSessionHandle()->RequiresTCPConnection()))
```
Again, there should be no checks for TCP here.

pidarped replied:
Yes, agreed. Removed.

bzbarsky-apple commented on src/messaging/ExchangeContext.cpp:
```diff
@@ -322,8 +323,11 @@ ExchangeContext::ExchangeContext(ExchangeManager * em, uint16_t ExchangeId, cons
-    SetAutoRequestAck(!session->IsGroupSession());
+    if (session->AllowsMRP())
+    {
+        // Do not request Ack for multicast
+        SetAutoRequestAck(!session->IsGroupSession());
```
This makes no sense now.  If AllowsMRP(), then IsGroupSession() is definitely false, right?

Fundamentally, we seem to be doing this thing where:

1. We `SetAutoRequestAck(!session->AllowsMRP())` here.
2. When we go to actually maybe request acks, we check `session->AllowsMRP()` again just in case someone did `SetAutoRequestAck(true)` on us in the meantime, I guess?  But then why do we bother checking the session for the initial value of "auto request ack"?  I guess it's OK, but seems to be confusing to both whoever wrote this code initially and you.

pidarped replied:
It is indeed a bit confusing in how all these separately named checks actually are directly mappable onto each other and so redundant.
But why `SetAutoRequestAck(!session->AllowsMRP());` ?
We actually need to do `SetAutoRequestAck(session->AllowsMRP());`. Did you mean that?

bzbarsky-apple replied:
>  But why SetAutoRequestAck(!session->AllowsMRP()); ?

Sorry, the `!` should not have been there.  It was late in the day....  

> I guess you meant that?

Yes. @pidarped

pidarped replied:
It is now `SetAutoRequestAck(session->AllowsMRP());`

bzbarsky-apple commented on src/messaging/ExchangeContext.cpp:
```diff
@@ -531,34 +535,49 @@ CHIP_ERROR ExchangeContext::HandleMessage(uint32_t messageCounter, const Payload
     {
-        if (!msgFlags.Has(MessageFlagValues::kDuplicateMessage) && payloadHeader.IsAckMsg() &&
-            payloadHeader.GetAckMessageCounter().HasValue())
+        if (mDispatch.IsReliableTransmissionAllowed() && !IsGroupExchangeContext())
```
If AllowsMRP(), then IsGroupExchangeContext() is false.

pidarped replied:
The whole check of `if (mDispatch.IsReliableTransmissionAllowed() && !IsGroupExchangeContext())` can be removed it seems.

bzbarsky-apple replied:
No, `mDispatch.IsReliableTransmissionAllowed()` needs to stay.  That's actually there for some test stuff that overrides default behavior.

pidarped replied:
It is there. I have removed the `!IsGroupExchangeContext()` part.

bzbarsky-apple commented on src/messaging/ExchangeContext.cpp:
```diff
@@ -531,34 +535,49 @@ CHIP_ERROR ExchangeContext::HandleMessage(uint32_t messageCounter, const Payload
-    {
-        return CHIP_NO_ERROR;
-    }
+        if (IsWaitingForAck())
```
This is wrong.  If we support MRP, and the incoming message was a duplicate that was not the ack for our current outstanding needs-to-be-acked message, that's not an error situation and we should not log it as such or return error.

In other words, this bit used to be after the duplicate check, and it needs to stay after the duplicate check.  Probably in a separate block conditioned on `AllowsMRP()`.

pidarped replied:
Yes, bringing it out of the check changed the order.

bzbarsky-apple commented on src/messaging/ExchangeMessageDispatch.cpp:
```diff
@@ -51,45 +51,60 @@ CHIP_ERROR ExchangeMessageDispatch::SendMessage(SessionManager * sessionManager,
+            payloadHeader.SetAckMessageCounter(reliableMessageContext->TakePendingPeerAckMessageCounter());
+        }
+
+        if (IsReliableTransmissionAllowed() && reliableMessageContext->AutoRequestAck() &&
+            reliableMessageContext->GetReliableMessageMgr() != nullptr && isReliableTransmission)
```
This is also pre-existing, but if we get here and `isReliableTransmission`, then `reliableMessageContext->AutoRequestAck()` is in fact true.  And `AllowsMRP()` is true.

So this seems like it would be much simpler if the `AllowsMRP()` check were just around the piggyback ack thing.  And if you really want, inside this block where isReliableTransmission is true you could VerifyOrDie(AllowsMRP()) or some equivalent, because that boolean must never be true if !AllowsMRP().

And then you would not need any other changes to this code at all.

pidarped replied:
It turns out IsReliableTransmissionAllowed() is hardcoded to true. Might be the same as AllowsMRP() FWIW and makes it more confusing. 
But my point in this change was to make connection-oriented flows bypass all these logic and not wade through them at all. So, I have encapsulated all these MRP specific changes within AllowsMRP() which TCP and BLE should bypass. That, I think, keeps the code cleaner and simpler I guess, no? Ultimately, code-change size wise, it might turn out to be almost similar.

bzbarsky-apple replied:
> It turns out IsReliableTransmissionAllowed() is hardcoded to true

It's a virtual method.  Some test bits subclass it and return false.

bzbarsky-apple commented on src/messaging/ExchangeContext.cpp:
```diff
@@ -142,7 +142,8 @@ CHIP_ERROR ExchangeContext::SendMessage(Protocols::Id protocolId, uint8_t msgTyp
 
     {
-        // ExchangeContext for group are supposed to always be Initiator
+        // ExchangeContext for group are supposed to always be Initiator AND not
+        // use TCP as the underlying transport.
```
This comment change no longer seems relevant....

bzbarsky-apple commented on src/messaging/ExchangeContext.cpp:
```diff
@@ -322,8 +323,8 @@ ExchangeContext::ExchangeContext(ExchangeManager * em, uint16_t ExchangeId, cons
 
-    // Do not request Ack for multicast
-    SetAutoRequestAck(!session->IsGroupSession());
+    // Request Ack when MRP is allowed. For group and connection-oriented sessions, this should be not set.
```
```suggestion
    // Try to use MRP by default, if it's possible over this session.
```

pidarped replied:
Changing it to "Try to use MRP by default, if it is allowed over this session."

bzbarsky-apple commented on src/messaging/ExchangeContext.cpp:
```diff
@@ -531,58 +532,65 @@ CHIP_ERROR ExchangeContext::HandleMessage(uint32_t messageCounter, const Payload
     });
 
-    if (mDispatch.IsReliableTransmissionAllowed() && !IsGroupExchangeContext())
+    if (GetSessionHandle()->AllowsMRP())
```
Is there a reason this is not just `if (mSession->AllowsMRP())`?  Seems like it should be the same thing in the end, but clearer...  Same for the condition below.

pidarped replied:
No particular reason. We could use `mSession` as well. Will change.

bzbarsky-apple commented on src/messaging/ExchangeMessageDispatch.cpp:
```diff
@@ -51,45 +51,60 @@ CHIP_ERROR ExchangeMessageDispatch::SendMessage(SessionManager * sessionManager,
     return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR ExchangeMessageDispatch::PrepareAndSendMessage(SessionManager * sessionManager, const SessionHandle & session,
```
Still not convinced all this surgery in this file is needed, but if we do it we should be clear about what sort of message is being sent here:
```suggestion
CHIP_ERROR ExchangeMessageDispatch::SendNonMRPMessage(SessionManager * sessionManager, const SessionHandle & session,
```

pidarped replied:
Changed it to PrepareAndSendNonMRPMessage

bzbarsky-apple commented on src/messaging/ExchangeMgr.cpp:
```diff
@@ -372,8 +372,10 @@ void ExchangeManager::SendStandaloneAckIfNeeded(const PacketHeader & packetHeade
-    // If we need to send a StandaloneAck, create a EphemeralExchange for the purpose to send the StandaloneAck
-    if (!payloadHeader.NeedsAck())
+
+    // If using the MRP protocol and we need to send a StandaloneAck, create a EphemeralExchange for the purpose to send
```
```suggestion
    // If using the MRP protocol and we need to send a StandaloneAck, create a EphemeralExchange to send
```

pidarped replied:
Noticed another typo in the original comment as well. `create an EphemeralExchange..`. Going to add that change as well.

pidarped committed:
Ensure that MRP logic is not executed for messages over other Transports

MRP based Ack handling and retransmissions should not be performed when
messages are not sent using MRP, e.g., over TCP.
Install guards on Send/Receive paths to steer traffic to bypass MRP
logic when going over TCP.

andy31415 commented on src/transport/GroupSession.h:
```diff
@@ -57,7 +57,8 @@ class IncomingGroupSession : public Session, public ReferenceCounted<IncomingGro
 
-    bool RequireMRP() const override { return false; }
+    bool AllowsMRP() const override { return false; }
+    bool IsTransportTCP() const override { return false; }
```
IsTransportTCP does not seem used but is defined. Could we drop it from this PR? We seem to only need `AllowsMRP`

pidarped replied:
The PeerAddress being of type TCP is one of the key flags that the underlying session establishment code would use to Connect over TCP, or use MRP. This function essentially returns that,(similar to what AllowsMRP does). AllowsMRP() is checking for UDP transport. So, essentially, it is the same as a IsTransportUDP(), isn't it? We need to associate the TCP connection with the session as we are not sharing sessions between MRP and TCP. Although this API is not being used in this PR, it would be used in subsequent changes.

andy31415 replied:
In particular this PR does not make use of IsTransportTCP so I have a hard time figuring out naming or usage suggestions.

I would be leaning towards a `SupportsLargePayloads` or `GetMaxPayloadSize` and `CloseSessionOnDisconnect` or similar items instead of a `if tcp` however I cannot check if my suggestions are sane without seeing usages of this IsTransportTCP. So I am requesting to remove this method and add it if really needed and used later on (and in those case to strongly consider similar naming of `Allow*` without passing in transport types up the chain outside transport.

pidarped replied:
@andy31415, renamed the IsTransportTCP() to AllowsLargePayload().

andy31415 reviewed (changes requested):
Please drop `IsTransportTCP` for now, we can add it when it is used.

I am looking to delay decisions that are transport specific until this is truly needed. For example I like `AllowMRP` a lot more than `!IsTCP` or `IsUDP` and am hoping we can make our interfaces present this kind of logic instead of transport-specific ones (like `AllowLargePackets`, `requireReconnect` or whatever reasoning we add later).

pidarped committed:
Rename IsTransportTCP() to AllowsLargePayload()
